File: packages/gather_events/lib/src/presence.dart
/// Current-state models: what the bridge believes the world looks like right
/// now. Sent as a snapshot when a client connects, so the app does not have to
/// replay history to render a first frame.
library;

/// One other person in the space, as far as the bridge can tell.
class PlayerRef {
  const PlayerRef({
    required this.id,
    this.name,
    this.inSpace = true,
    this.isFollowingMe = false,
    this.speaking = false,
    this.micOn,
    this.cameraOn,
    this.screensharing = false,
    this.followingMeSince,
  });

  /// Gather's player uuid. Stable for the duration of a session.
  final String id;

  /// Display name, straight from Gather's `SpaceUser.name`.
  final String? name;

  final bool inSpace;

  /// This person is following me around — the one thing worth knowing about
  /// somebody else in the space, and the reason this app exists.
  final bool isFollowingMe;

  /// Talking right now — Gather's own `SpaceUser.speaking`, and the most
  /// frequently updated field on the whole game socket. The bridge only reports
  /// changes for people who are following you, because they are the only ones
  /// the app draws.
  final bool speaking;

  /// Always null. Mic, camera and screenshare were only ever readable by
  /// scraping the desktop client's log and are in no Gather model; [speaking]
  /// is the signal that replaced them. Kept so older snapshots still decode.
  final bool? micOn;
  final bool? cameraOn;
  final bool screensharing;

  final DateTime? followingMeSince;

  /// What to show in a list: the name if we have it, else a short id.
  String get label => name ?? id.substring(0, id.length.clamp(0, 8));

  PlayerRef copyWith({
    String? name,
    bool? inSpace,
    bool? isFollowingMe,
    bool? speaking,
    bool? micOn,
    bool? cameraOn,
    bool? screensharing,
    DateTime? followingMeSince,
    bool clearFollowingMeSince = false,
  }) {
    return PlayerRef(
      id: id,
      name: name ?? this.name,
      inSpace: inSpace ?? this.inSpace,
      isFollowingMe: isFollowingMe ?? this.isFollowingMe,
      speaking: speaking ?? this.speaking,
      micOn: micOn ?? this.micOn,
      cameraOn: cameraOn ?? this.cameraOn,
      screensharing: screensharing ?? this.screensharing,
      followingMeSince: clearFollowingMeSince
          ? null
          : (followingMeSince ?? this.followingMeSince),
    );
  }

  Map<String, Object?> toJson() => {
        'id': id,
        'name': name,
        'inSpace': inSpace,
        'isFollowingMe': isFollowingMe,
        'speaking': speaking,
        'micOn': micOn,
        'cameraOn': cameraOn,
        'screensharing': screensharing,
        'followingMeSince': followingMeSince?.toIso8601String(),
      };

  static PlayerRef fromJson(Map<String, Object?> json) => PlayerRef(
        id: json['id'] as String? ?? '',
        name: json['name'] as String?,
        inSpace: json['inSpace'] as bool? ?? true,
        isFollowingMe: json['isFollowingMe'] as bool? ?? false,
        speaking: json['speaking'] as bool? ?? false,
        micOn: json['micOn'] as bool?,
        cameraOn: json['cameraOn'] as bool?,
        screensharing: json['screensharing'] as bool? ?? false,
        followingMeSince:
            DateTime.tryParse(json['followingMeSince'] as String? ?? ''),
      );
}

/// My own state in the space.
class SelfState {
  const SelfState({
    this.userId,
    this.spaceId,
    this.spaceName,
    this.micOn,
    this.cameraOn,
    this.screensharing = false,
    this.inOffice,
    this.followingPlayerId,
  });

  final String? userId;
  final String? spaceId;
  final String? spaceName;
  final bool? micOn;
  final bool? cameraOn;
  final bool screensharing;
  final bool? inOffice;

  /// Who I am currently following, if anyone.
  final String? followingPlayerId;

  SelfState copyWith({
    String? userId,
    String? spaceId,
    String? spaceName,
    bool? micOn,
    bool? cameraOn,
    bool? screensharing,
    bool? inOffice,
    String? followingPlayerId,
    bool clearFollowing = false,
  }) {
    return SelfState(
      userId: userId ?? this.userId,
      spaceId: spaceId ?? this.spaceId,
      spaceName: spaceName ?? this.spaceName,
      micOn: micOn ?? this.micOn,
      cameraOn: cameraOn ?? this.cameraOn,
      screensharing: screensharing ?? this.screensharing,
      inOffice: inOffice ?? this.inOffice,
      followingPlayerId:
          clearFollowing ? null : (followingPlayerId ?? this.followingPlayerId),
    );
  }

  Map<String, Object?> toJson() => {
        'userId': userId,
        'spaceId': spaceId,
        'spaceName': spaceName,
        'micOn': micOn,
        'cameraOn': cameraOn,
        'screensharing': screensharing,
        'inOffice': inOffice,
        'followingPlayerId': followingPlayerId,
      };

  static SelfState fromJson(Map<String, Object?> json) => SelfState(
        userId: json['userId'] as String?,
        spaceId: json['spaceId'] as String?,
        spaceName: json['spaceName'] as String?,
        micOn: json['micOn'] as bool?,
        cameraOn: json['cameraOn'] as bool?,
        screensharing: json['screensharing'] as bool? ?? false,
        inOffice: json['inOffice'] as bool?,
        followingPlayerId: json['followingPlayerId'] as String?,
      );
}

/// What the bridge is currently connected to, so the app can show honestly how
/// good its information is.
class CollectorHealth {
  const CollectorHealth({
    this.gather = false,
    this.logTail = false,
    this.cdp = false,
    this.detail,
  });

  /// The connection to Gather's own game socket. Everything about people —
  /// names, following, voice activity — comes from here.
  final bool gather;

  /// The tail of the desktop client's log, which now carries exactly one thing:
  /// Gather's own notifications (waves, meeting invites, event reminders). Down
  /// simply means the desktop app is not running; presence is unaffected.
  final bool logTail;

  /// Was the CDP collector, which no longer exists. Still read because a bridge
  /// older than this app build sets it and nothing else, and because the current
  /// bridge mirrors [gather] onto it for app builds older than *it*.
  final bool cdp;

  final String? detail;

  /// Whether we have names, coordinates and explicit follow state.
  bool get hasRichData => gather || cdp;

  Map<String, Object?> toJson() => {
        'gather': gather,
        'logTail': logTail,
        'cdp': cdp,
        'detail': detail,
      };

  static CollectorHealth fromJson(Map<String, Object?> json) => CollectorHealth(
        gather: json['gather'] as bool? ?? false,
        logTail: json['logTail'] as bool? ?? false,
        cdp: json['cdp'] as bool? ?? false,
        detail: json['detail'] as String?,
      );
}

/// Party mode: whether the bridge is currently teleporting me around the map.
///
/// Comes back on the snapshot rather than being tracked in the app, because the
/// app is not the only thing that can change it — the bridge stops on its own
/// after fifteen minutes, when it loses Gather, and when the daemon shuts down.
/// A toggle that only remembers what it was last told would keep glowing through
/// all three.
class PartyState {
  const PartyState({
    this.active = false,
    this.hops = 0,
    this.safeTiles = 0,
    this.detail,
  });

  final bool active;

  /// Teleports made this session.
  final int hops;

  /// How many tiles were far enough from everyone at the last hop.
  final int safeTiles;

  /// Why it is not hopping, when it is not. Also carries the reason it stopped.
  final String? detail;

  Map<String, Object?> toJson() => {
        'active': active,
        'hops': hops,
        'safeTiles': safeTiles,
        'detail': detail,
      };

  static PartyState fromJson(Map<String, Object?> json) => PartyState(
        active: json['active'] as bool? ?? false,
        hops: (json['hops'] as num?)?.toInt() ?? 0,
        safeTiles: (json['safeTiles'] as num?)?.toInt() ?? 0,
        detail: json['detail'] as String?,
      );
}

/// Full snapshot pushed on connect and whenever state changes materially.
class PresenceSnapshot {
  const PresenceSnapshot({
    required this.self,
    required this.players,
    required this.health,
    required this.at,
    this.party = const PartyState(),
  });

  final SelfState self;
  final List<PlayerRef> players;
  final CollectorHealth health;
  final DateTime at;

  /// Defaulted rather than required: a bridge older than this app build sends no
  /// `party` key at all, and "off" is the right reading of its silence.
  final PartyState party;

  List<PlayerRef> get followers =>
      players.where((p) => p.isFollowingMe).toList();

  /// The same world, with party mode's counters moved on.
  ///
  /// The bridge sends the hop counter on its own small frame rather than
  /// republishing the roster for it — see `BridgeServer._publishParty`. This is
  /// how that frame is folded in.
  PresenceSnapshot withParty(PartyState party) => PresenceSnapshot(
        self: self,
        players: players,
        health: health,
        at: at,
        party: party,
      );

  Map<String, Object?> toJson() => {
        'type': 'presence.snapshot',
        'at': at.toIso8601String(),
        'self': self.toJson(),
        'players': players.map((p) => p.toJson()).toList(),
        'health': health.toJson(),
        'party': party.toJson(),
      };

  static PresenceSnapshot fromJson(Map<String, Object?> json) =>
      PresenceSnapshot(
        at: DateTime.tryParse(json['at'] as String? ?? '') ?? DateTime.now(),
        self: SelfState.fromJson(
            (json['self'] as Map?)?.cast<String, Object?>() ?? const {}),
        players: ((json['players'] as List?) ?? const [])
            .map((p) => PlayerRef.fromJson((p as Map).cast<String, Object?>()))
            .toList(),
        health: CollectorHealth.fromJson(
            (json['health'] as Map?)?.cast<String, Object?>() ?? const {}),
        party: PartyState.fromJson(
            (json['party'] as Map?)?.cast<String, Object?>() ?? const {}),
      );

  /// A blank snapshot for first paint, before the bridge has answered.
  static PresenceSnapshot get empty => PresenceSnapshot(
        self: const SelfState(),
        players: const [],
        health: const CollectorHealth(),
        at: DateTime.fromMillisecondsSinceEpoch(0),
      );
}

File: bridge/lib/presence.js
/**
 * Folds the roster into "what does the world look like right now".
 *
 * There used to be two fidelities of input here: inferred log events and observed
 * `SpaceUser` rows. Only the second remains. Everything now arrives as Gather's
 * own model rows from `DirectCollector`, so proximity comes from `clusterId`
 * (Gather's own notion of "standing together") and following from
 * `followTargetId`, and both are *observed* rather than guessed.
 *
 * What that removed, and why it is not missed:
 *
 *  - The movement-based follow detector. It existed because the log had no
 *    `followTargetId` to read. With the real field always present, guessing from
 *    convergence is strictly worse — and it had already stopped being reachable.
 *  - Per-player mic/camera/screenshare. Those came from the client's log and are
 *    in no Gather model at all. `speaking` — live voice activity — is in the game
 *    state and is the better signal anyway: it says who is *talking*, not who
 *    happens to have a mic enabled.
 *
 * One log-derived event still passes through `apply`: `notification.shown`, for
 * the waves and meeting invites Gather's own client raises. It needs no folding —
 * it is news, not state — so it falls through to the default branch.
 */

import { bridgeStatus, emptySelf, follow, newPlayer, playerMoved, proximity } from './events.js';

/** How close, in tiles, counts as "standing next to me". */
export const ADJACENT_TILES = 3;

/**
 * Someone already counted as near stays near until they pass this. The gap
 * between the two is deliberate: positions arrive debounced, so a single
 * threshold makes anyone loitering at exactly the boundary flap between
 * "arrived" and "left" indefinitely — and every flap is a notification.
 */
export const LEAVE_TILES = 4.5;

export class PresenceTracker {
  constructor() {
    /** @type {Map<string, object>} */
    this._players = new Map();
    this._self = emptySelf();
    this._health = { gather: false, cdp: false, logTail: false, detail: null };
    /**
     * Party mode, mirrored here so it rides along on the snapshot the phone
     * already receives. It is the one piece of state the app can *change* rather
     * than only observe, which makes reporting it back the difference between a
     * button that knows what it did and one that only hopes.
     */
    this._party = { active: false, hops: 0, safeTiles: 0, detail: null };
    this._selfPosition = null;
    /** My own SpaceUser id and cluster, learned from the roster. */
    this._selfId = null;
    this._selfClusterId = null;
    this._selfFloorId = null;
  }

  get self() {
    return this._self;
  }

  get health() {
    return this._health;
  }

  get players() {
    return [...this._players.values()];
  }

  get party() {
    return this._party;
  }

  snapshot() {
    return {
      type: 'presence.snapshot',
      at: new Date().toISOString(),
      self: this._self,
      players: this.players,
      health: this._health,
      party: this._party,
    };
  }

  setHealth(patch) {
    this._health = { ...this._health, ...patch };
  }

  setParty(state) {
    this._party = { ...this._party, ...state };
  }

  _player(id) {
    let p = this._players.get(id);
    if (!p) {
      p = newPlayer(id);
      this._players.set(id, p);
    }
    return p;
  }

  /**
   * Apply one event the collector produced directly, rather than via the roster.
   *
   * That is now a short list — status and the space the bridge was told to watch.
   * Everything about *people* comes through `applyRoster`.
   *
   * Returns `{ emit, stateChanged }` where `emit` holds only the events worth
   * showing a human.
   */
  apply(event) {
    switch (event.type) {
      case 'self.changed': {
        const before = { ...this._self };
        if (event.userId != null) this._self.userId = event.userId;
        if (event.inOffice != null) this._self.inOffice = event.inOffice;
        const changed =
          before.userId !== this._self.userId || before.inOffice !== this._self.inOffice;
        if (!changed) return { emit: [], stateChanged: false };
        return { emit: [event], stateChanged: true };
      }

      case 'space.changed': {
        const changed = event.spaceId != null && event.spaceId !== this._self.spaceId;
        if (event.spaceId != null) this._self.spaceId = event.spaceId;
        if (event.spaceName != null) this._self.spaceName = event.spaceName;
        if (changed) this._clearRoom();
        return { emit: [event], stateChanged: true };
      }

      case 'bridge.status': {
        const patch = {
          [event.collector]: event.healthy,
          detail: event.detail ?? this._health.detail,
        };
        // `cdp` is a compatibility alias, not a second collector. App builds
        // already on people's phones compute `hasRichData` from
        // `CollectorHealth.cdp`, so publishing only the honest name would make
        // them show the "log-only mode: no names" banner while we hold the full
        // roster. Newer builds read `gather` and ignore this.
        if (event.collector === 'gather') patch.cdp = event.healthy;
        this.setHealth(patch);
        return { emit: [event], stateChanged: true };
      }

      default:
        return { emit: [event], stateChanged: false };
    }
  }

  _clearRoom() {
    this._players.clear();
    this._selfClusterId = null;
  }

  /**
   * Replace what we know from Gather's own `SpaceUser` records.
   *
   * @param {{ selfId: string|null, rows: Array<object>, spaceName?: string|null }} roster
   * @returns {{ emit: object[], stateChanged: boolean }}
   */
  applyRoster({ selfId, rows, spaceName = null }) {
    const emit = [];
    let stateChanged = false;
    const at = new Date();

    if (spaceName && spaceName !== this._self.spaceName) {
      this._self.spaceName = spaceName;
      stateChanged = true;
    }

    if (selfId) this._selfId = selfId;
    const me = rows.find((r) => r.id === this._selfId) ?? null;

    // Whether *you* are in the space at all. Our own connection is an observer —
    // it never calls `enterSpace` — so this reflects your real clients, not us.
    //
    // It matters because your avatar does not disappear when you close Gather: it
    // stays parked wherever you left it, usually at your desk. Judging proximity
    // against a parked avatar reports every passer-by as standing next to you, all
    // night. So when you are offline, nobody is near you, by definition.
    const selfOnline = me?.connected !== false;
    if (me && this._self.inOffice !== selfOnline) {
      this._self.inOffice = selfOnline;
      stateChanged = true;
    }

    if (me) {
      this._selfClusterId = me.clusterId ?? null;
      this._selfFloorId = me.floorId ?? null;
      if (me.x != null && me.y != null) this._selfPosition = [me.x, me.y];
      if (me.followTargetId !== undefined) {
        const wanted = me.followTargetId ?? null;
        if (wanted !== this._self.followingPlayerId) {
          this._self.followingPlayerId = wanted;
          stateChanged = true;
        }
      }
    }

    // Without knowing which row is *me*, "next to me" and "following me" are both
    // undefined. Merge the identity information the roster does give us — names
    // and coordinates are still worth having — and judge nothing.
    const knowSelf = this._selfId != null;

    const seen = new Set();

    for (const row of rows) {
      if (!row?.id || row.id === this._selfId) continue;
      seen.add(row.id);
      const p = this._player(row.id);

      if (row.name && row.name !== p.name) {
        p.name = row.name;
        stateChanged = true;
      }
      if (row.speaking != null && row.speaking !== p.speaking) {
        // State only, never an event: voice activity toggles every few seconds
        // while someone talks, and a feed of that is unreadable.
        //
        // It is also only a *state change* for somebody the app draws. The app
        // shows the talking indicator on the people standing next to you and the
        // ones following you, and on nobody else — but the measured space holds 79
        // rows, so counting every one of them made a stranger three rooms away
        // clearing their throat publish a 23 KiB roster to every phone. That was
        // the bulk of an observed 0.86 snapshots/second, ~1.1 MB/minute, which is
        // what made the link feel fragile: the phone spent its time decoding
        // rosters and rebuilding its whole widget tree. Whoever is invisible keeps
        // the new value on the player record, so it is already correct in the
        // snapshot that goes out when they *do* walk up.
        const visible = p.isNear || p.isFollowingMe;
        p.speaking = row.speaking;
        if (visible) stateChanged = true;
      }
      // Whether they actually changed tiles, decided before the row overwrites
      // what we held. A first sighting is not a move — on the initial state dump
      // every player would otherwise "move" from nowhere to where they stand.
      const hadPosition = p.x != null && p.y != null;
      const moved =
        hadPosition && ((row.x != null && row.x !== p.x) || (row.y != null && row.y !== p.y));

      if (row.x != null) p.x = row.x;
      if (row.y != null) p.y = row.y;

      const distance =
        this._selfPosition && row.x != null && row.y != null
          ? Math.hypot(row.x - this._selfPosition[0], row.y - this._selfPosition[1])
          : null;
      // Rounded for the wire only. The full-precision `distance` is what the
      // near/far thresholds below compare against, so this cannot move anyone
      // across a boundary — it just stops `19.4164878389476` costing 19 bytes per
      // player per snapshot to say "about 19 tiles away".
      if (distance != null) p.distance = Math.round(distance * 10) / 10;

      // A row that is not connected is a stale avatar left wherever that person
      // logged off — which, for most people, is their desk. It keeps its name and
      // its coordinates, so on distance alone it is indistinguishable from
      // somebody standing next to you, and walking past an empty desk reports its
      // owner as having walked up. The full state dump carries every member of
      // the space, not just the ones online: 111 rows in the measured space, most
      // of them offline and still holding coordinates. Only the connected ones can
      // be standing anywhere. `null` stays judgeable — unknown is not absent.
      const online = row.connected !== false;

      // Gather's own proximity predicate requires the same floor before any
      // distance comparison — two people on the same tile of different floors are
      // not near each other. Treat unknown floors as "no objection".
      const sameFloor =
        this._selfFloorId == null || row.floorId == null || row.floorId === this._selfFloorId;

      // Adjacency: sharing a cluster is Gather's own notion of "we are standing
      // together". Distance decides when clusters are unavailable, with a wider
      // radius for leaving than for arriving.
      const sameCluster =
        this._selfClusterId != null && row.clusterId != null && row.clusterId === this._selfClusterId;
      const closeEnough =
        distance != null && distance <= (p.isNear ? LEAVE_TILES : ADJACENT_TILES);
      const near = selfOnline && online && sameFloor && (sameCluster || closeEnough);

      // With neither a position nor a cluster there is nothing to judge on. Say
      // nothing rather than reporting everyone as having walked off. Going offline
      // is always judgeable: whatever we thought before, they are not standing
      // there now. So is our own absence.
      const canJudge = !selfOnline || !online || sameCluster || distance != null;

      // Someone next to you shifting around is worth saying out loud; the other
      // hundred rows changing tiles four times a second is not, and the snapshot
      // already carries every position for anyone who wants the map. So this is
      // deliberately scoped to the people the app is actually watching.
      if (moved && online && p.isNear) {
        emit.push(playerMoved({ at, playerId: row.id, x: p.x, y: p.y, distance }));
      }

      if (knowSelf && canJudge && near !== p.isNear) {
        p.isNear = near;
        p.nearSince = near ? at.toISOString() : null;
        stateChanged = true;
        emit.push(
          proximity({
            at,
            confidence: sameCluster ? 'observed' : 'inferred',
            playerId: row.id,
            near,
            distance,
          }),
        );
      }

      // Following: their followTargetId pointing at me is the real thing, no
      // heuristics needed.
      if (knowSelf && row.followTargetId !== undefined) {
        const followsMe = row.followTargetId === this._selfId;
        if (followsMe !== p.isFollowingMe) {
          p.isFollowingMe = followsMe;
          p.followingMeSince = followsMe ? at.toISOString() : null;
          stateChanged = true;
          emit.push(
            follow({
              at,
              followerId: row.id,
              targetId: this._selfId ?? 'self',
              started: followsMe,
              targetIsSelf: true,
            }),
          );
        }
      }

      if (row.connected != null && p.inSpace !== row.connected) {
        p.inSpace = row.connected;
        stateChanged = true;
      }
    }

    // Anyone previously in the roster but absent now has left.
    for (const [id, p] of this._players) {
      if (!knowSelf) break;
      if (seen.has(id) || !p.inSpace) continue;
      if (rows.length === 0) break; // an empty roster means "unknown", not "empty"
      p.inSpace = false;
      if (p.isNear) {
        p.isNear = false;
        p.nearSince = null;
        emit.push(proximity({ at, playerId: id, near: false }));
      }
      if (p.isFollowingMe) {
        p.isFollowingMe = false;
        p.followingMeSince = null;
        emit.push(
          follow({
            at,
            followerId: id,
            targetId: this._selfId ?? 'self',
            started: false,
            targetIsSelf: true,
          }),
        );
      }
      stateChanged = true;
    }

    return { emit, stateChanged };
  }

  statusEvent(collector, healthy, detail) {
    this.setHealth({ [collector]: healthy, detail: detail ?? null });
    return bridgeStatus({ at: new Date(), collector, healthy, detail });
  }
}
